[Hotfix] Docs links and API docs (Swagger UI) updates

suzinyou opened the pull request:
Reviewer: @Tanmay-97 @amiraliemami 
Estimate: 45min

---

## Ticket

Fixes: Addressing a few of @ziakhan04 's feedback

## Description

### Goal

1. Link to app and API docs from doc site
2. Improve API docs with descriptions and examples
3. Rename JSON fields for clarity ("score"-> "distance", "failed_rules" -> "matched_rules")

### Changes
1. Docs site now has two more menu items: Try it out (to app), API docs (to Swagger UI)
    <img width="969" alt="image" src="https://github.com/user-attachments/assets/3f1ba5cc-13b5-499e-8fcd-86b1ee42a4fb">
2. Improve Swagger UI
    1. Each router.py should now define `TAG_METADATA` for Swagger UI tag and description
    4. Improved pydantic docstring for some
    5. Added json examples for core APIs (content mgmt, question answering, UD rules, UD)
    <img width="936" alt="Screenshot 2024-07-17 at 1 09 54 AM" src="https://github.com/user-attachments/assets/78032943-0530-48ce-9d26-7653c5931e68">
3. Updated schema for clarity
    1. Content retrieval "score" -> "distance"
    2. Urgency detection "failed_rules" -> "matched_rules"
    3. Several other return type changes for clarity


### Future Tasks (optional)

1. ~**Add target="_blank" for the external navbar links in mkdocs**~ New plugin added!
4. Customize Swagger UI to include more descriptions, logo, links back to docs, etc. ([ticket](https://idinsight.atlassian.net/browse/AAQ-667))

## How has this been tested?

1. Check docs navbar links: 
    ```shell
    cd aaq-core
    mkdocs serve -a "localhost:8080"
    ```
2. Check Swagger UI:
    ```shell
    cd aaq-core/core_backend
    python main.py
    ```

## To-do before merge (optional)

- [x] Merge #299 and resolve conflicts

## Checklist

Fill with `x` for completed. 

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts

suzinyou commented on core_backend/app/__init__.py:
```diff
@@ -26,6 +26,16 @@
 
 logger = setup_logger()
 
+tags_metadata = [
+    contents.TAG_METADATA,
+    tags.TAG_METADATA,
+    question_answer.TAG_METADATA,
+    urgency_rules.TAG_METADATA,
+    urgency_detection.TAG_METADATA,
+    auth.TAG_METADATA,
+    dashboard.TAG_METADATA,
+    user_tools.TAG_METADATA,
+]
```
Each metadata defines tag name and description

suzinyou commented on core_backend/app/contents/routers.py:
```diff
@@ -58,7 +63,9 @@ async def create_content(
     """
     Create content endpoint. Calls embedding model to get content embedding and
-    inserts it to PG database
+    inserts it to the content table in the database.
+
+    ⚠️ To add tags, use the tags endpoint to create tags first.
```
These docstrings are rendered on Swagger UI as endpoint description, so made improvements here and there!

suzinyou commented on core_backend/app/question_answer/schemas.py:
```diff
@@ -9,18 +9,28 @@
 class QueryBase(BaseModel):
     """
-    Pydantic model for query APIs
+    Question answering query base class.
```
This is also rendered on Swagger UI (under Schemas section at the bottom)
Made changes wherever I saw fit

<img width="434" alt="image" src="https://github.com/user-attachments/assets/b85abc19-a7e8-4473-900d-7e3f5b59e0e8">

suzinyou commented on core_backend/app/contents/models.py:
Turned some ambiguous types into Pydantic classes. This is one example

amiraliemami replied:
god bless you

suzinyou commented on core_backend/app/question_answer/routers.py:
```diff
@@ -230,14 +235,12 @@ async def get_semantic_matches(
     """
     if not isinstance(response, QueryResponseError):
         metadata = create_langfuse_metadata(query_id=response.query_id, user_id=user_id)
-        content_response = convert_search_results_to_schema(
```
now get_similar_content_async directly returns the schema object, so this function is not necessary

suzinyou commented on core_backend/app/schemas.py:
```diff
@@ -21,6 +21,6 @@ class QuerySearchResult(BaseModel):
     retrieved_text: str
     retrieved_content_id: int
-    score: float
+    distance: float
```
Fixed the naming which had been confusing everyone!

amiraliemami replied:
ty!

suzinyou commented on core_backend/app/tags/schemas.py:
```diff
@@ -24,18 +36,23 @@ class TagRetrieve(TagCreate):
     updated_datetime_utc: datetime
 
-
-class TagUpdate(TagCreate):
-    """
-    Pydantic model for tag edit
-    """
-
-    content_id: int
-
-
-class TagDelete(BaseModel):
-    """
-    Pydantic model for tag deletion
-    """
-
-    content_id: int
```
These were not used anywhere.

amiraliemami commented on core_backend/app/llm_call/entailment.py:
```diff
@@ -12,7 +12,7 @@
 async def detect_urgency(
     urgency_rules: List[str], message: str, metadata: Optional[dict] = None
-) -> Dict[str, Any]:
+) -> UrgencyDetectionEntailment.UrgencyDetectionEntailmentResult:
```
nice

amiraliemami commented on core_backend/app/llm_call/entailment.py:
```diff
@@ -34,4 +34,4 @@ async def detect_urgency(
         parsed_json = ud_entailment.default_json
 
-    return parsed_json
+    return UrgencyDetectionEntailment.UrgencyDetectionEntailmentResult(**parsed_json)
```
nice

amiraliemami commented on core_backend/app/contents/models.py:
```diff
@@ -271,7 +271,7 @@ async def get_similar_content_async(
     asession: AsyncSession,
     metadata: Optional[dict] = None,
-) -> Dict[int, tuple[str, str, int, float]]:
+) -> Dict[int, QuerySearchResult]:
```
Nice

amiraliemami commented on core_backend/app/contents/models.py:
```diff
@@ -315,7 +315,12 @@ async def get_search_results(
     results_dict = {}
     for i, r in enumerate(search_result):
-        results_dict[i] = (r[0].content_title, r[0].content_text, r[0].content_id, r[1])
+        results_dict[i] = QuerySearchResult(
+            retrieved_content_id=r[0].content_id,
+            retrieved_title=r[0].content_title,
+            retrieved_text=r[0].content_text,
+            distance=r[1],
+        )
```
this is bliss compared to before

amiraliemami commented on core_backend/app/contents/schemas.py:
```diff
@@ -15,12 +15,22 @@ class ContentCreate(BaseModel):
+        from_attributes=True,
+        json_schema_extra={
+            "examples": [
+                {
+                    "content_title": "An example content title",
+                    "content_text": "And an example content text!",
+                },
```
for some reason the order of these are reversed on the UI which make it a little confusing... any fix there?

suzinyou replied:
Yeah I noticed that, but I'm not sure :(

amiraliemami replied:
I think this is kinda crucial for easy comprehension... I would actually prefer to not have examples but have them the right way round

amiraliemami commented on core_backend/app/question_answer/schemas.py:
```diff
@@ -9,18 +9,28 @@
+        json_schema_extra={
+            "examples": [
+                {
+                    "query_text": "What is AAQ?",
+                    "query_metadata": {"source": "optinal-metadata-example"},
```
these also show up flipped... we really want the query text to come up first, could you look into it?

amiraliemami commented on core_backend/app/question_answer/schemas.py:
```diff
@@ -63,7 +73,35 @@ class QueryResponse(BaseModel):
 
-    model_config = ConfigDict(from_attributes=True)
+    model_config = ConfigDict(
+        from_attributes=True,
```
beautiful

amiraliemami commented on core_backend/app/urgency_rules/schemas.py:
```diff
@@ -24,3 +36,34 @@ class UrgencyRuleRetrieve(UrgencyRuleCreate):
+    )
+
+
+class UrgencyRuleCosineDistance(BaseModel):
```
can you add a docstring to this pls

amiraliemami commented on core_backend/app/__init__.py:
```diff
@@ -26,6 +26,35 @@
+application.
+- **Urgency rules management**: APIs to manage the urgency rules in the application.
+"""
+tags_metadata = [
+    question_answer.TAG_METADATA,
+    urgency_detection.TAG_METADATA,
+    contents.TAG_METADATA,
+    tags.TAG_METADATA,
+    urgency_rules.TAG_METADATA,
+    dashboard.TAG_METADATA,
+    auth.TAG_METADATA,
+    user_tools.TAG_METADATA,
+    admin.TAG_METADATA,
+]
```
I reordered these to reflect people's usecase when they come to this page. also corresponds with description above. what do you think?

suzinyou replied:
Sure ! Thank you for the page description also, I didn't know about this param!

I had put content first because question_answering is not useful without content. But this ordering also makes sense!

amiraliemami commented on core_backend/app/question_answer/schemas.py:
```diff
@@ -63,7 +73,35 @@ class QueryResponse(BaseModel):
+    model_config = ConfigDict(
+        from_attributes=True,
+        json_schema_extra={
+            "examples": [
+                {
+                    "query_id": 1,
+                    "content_response": {
+                        "0": {
+                            "retrieved_title": "Example content title",
+                            "retrieved_text": "Example content text",
+                            "retrieved_content_id": 23,
+                            "distance": 0.1,
+                        },
+                        "1": {
+                            "retrieved_title": "Another example content title",
+                            "retrieved_text": "Another example content text",
+                            "retrieved_content_id": 12,
+                            "distance": 0.2,
+                        },
+                    },
+                    "llm_response": "Example LLM response "
+                    "(null for /embeddings-search)",
+                    "feedback_secret_key": "secret-key-12345-abcde",
+                    "debug_info": {"example": "debug-info"},
+                    "state": "final",
+                }
+            ]
+        },
+    )
```
honestly I'm a bit worried all these examples, as nice as they are, are gonna end up being another thing we need to update when we change something... like we'll have to add this to our PR checklist. The auto-generated stuff works well enough in my opinion, but open to adding these if you really see value in them. Maybe we can ask the rest of the team too?

suzinyou replied:
Good point. But mypy was cathing errors if the example didn't align with the schema, so
perhaps it won't be too bad? I was thinking of Zia's feedback about how it's not clear
what to put where. I think giving concrete examples contributes to making AAQ more dev
friendly.. so feel like it's a worthwhile thing to maintain!

amiraliemami replied:
ok, if you think it's worth maintaining then let's do it!

amiraliemami commented on core_backend/app/contents/routers.py:
```diff
@@ -30,7 +30,13 @@
 )
 
-router = APIRouter(prefix="/content", tags=["Content Management"])
+TAG_METADATA = {
+    "name": "Question-answering content management",
+    "description": "Authentication: User login. Manage content and tags to be used for "
+    "question answering",
+}
```
Thoughts on dropping the "Question-answering" part of this and the tags one? Makes the line super long which makes the description squish up against it badly when you resize the window to be small. Or we can rethink the inclusion of the descriptions now that we have the little blurb at the top of the page?

suzinyou replied:
Cool with it! I'll update

amiraliemami reviewed (commented):
Thanks for all this! Lots of great fixes.

I am a little unsure about two things this PR introduces:
- payload examples: these might become cumbersome to maintain? the autogenerated ones were good enough in my opinion but I also see that these are definitely better (esp. if we can figure out the ordering issue)
- API router descriptions: they squish up really badly when the window is made smaller... Thoughts on abandoning them? I added a little description to the top of the page so that might be enough? Or perhaps just keep them for explicitly saying the Auth type.

<img width="575" alt="image" src="https://github.com/user-attachments/assets/5d77227c-0d30-49a0-8c1c-1d463b70d167">

suzinyou commented:
> * payload examples: these might become cumbersome to maintain? the autogenerated ones were good enough in my opinion but I also see that these are definitely better (esp. if we can figure out the ordering issue)
> * API router descriptions: they squish up really badly when the window is made smaller... Thoughts on abandoning them? I added a little description to the top of the page so that might be enough? Or perhaps just keep them for explicitly saying the Auth type.

Payload: I've responded in the other comment but I think it's worth it, since mypy will catch errors if you don't update it

Second point, I don't think the primary use case is for mobile, so not too worried about this. Wdyt?

Also thank you so much for all the edits ! I added some links to the swaggerui description!

suzinyou commented on core_backend/app/__init__.py:
```diff
@@ -26,6 +26,58 @@
+🔑 <a href="https://app.ask-a-question.com/integrations" target="_blank">Get your API
+key</a> |
+📖 <a href="https://idinsight.github.io/aaq-core" target="_blank">Docs</a> |
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" height="1em"
+style="vertical-align: bottom;"><path d="m0 0h24v24h-24z" fill="#fff" opacity="0"
+transform="matrix(-1 0 0 -1 24 24)"/><path d="m12 1a10.89 10.89 0 0 0 -11 10.77 10.79
+10.79 0 0 0 7.52 10.23c.55.1.75-.23.75-.52s0-.93
+0-1.83c-3.06.65-3.71-1.44-3.71-1.44a2.86 2.86 0 0 0 -1.22-1.58c-1-.66.08-.65.08-.65a2.31
+2.31 0 0 1 1.68 1.11 2.37 2.37 0 0 0 3.2.89 2.33 2.33 0 0 1
+.7-1.44c-2.44-.27-5-1.19-5-5.32a4.15 4.15 0 0 1 1.11-2.91 3.78 3.78 0 0 1
+.11-2.84s.93-.29 3 1.1a10.68 10.68 0 0 1 5.5 0c2.1-1.39 3-1.1 3-1.1a3.78 3.78 0 0 1 .11
+2.84 4.15 4.15 0 0 1 1.17 2.89c0 4.14-2.58 5.05-5 5.32a2.5 2.5 0 0 1 .75
+2v2.95s.2.63.75.52a10.8 10.8 0 0 0 7.5-10.22 10.89 10.89 0 0 0 -11-10.77"
+fill="#231f20"/></svg> <a href="https://github.com/IDinsight/aaq-core"
+target="_blank">GitHub</a> |
```
Happy to replace with an appropriate emoji...

amiraliemami reviewed (approved):
Makes sense, thanks!
